### PR TITLE
feature: Enable strict mode for workspace search #2032

### LINKED ISSUE #2032: Enable strict mode for workspace-search
### Check for duplicates

- [X] I have searched for similar issues before opening a new one.

### Component

workspace-search

### Problem

_No response_

### Request

The workspace-search plugin does not have "[strict mode](https://www.typescriptlang.org/tsconfig#strict)" enabled for TypeScript. We should enable it to have better type checking for our plugins, but doing so may show some TypeScript errors that need to be fixed first.

This would be a good first issue for anyone experienced with working in TypeScript. You may need to adjust types, add checks for null or undefined, or make other fixes. If you have questions, feel free to ask here! In general, the team prefers avoiding casts and using `!` to coerce non-null types, and instead adding appropriate checks.

To work on this issue:

1. Run `npm install` at the root of Blockly samples to set up dependencies.
2. Navigate to the plugin's directory
1. Change `strict: false` to `strict: true` in the plugin's tsconfig.json
2. Run `npm run build` and see if there are any TypeScript errors.
3. Resolve any TypeScript errors, if present. If not, then we got lucky and accidentally wrote strict code! Yay!
4. Run `npm run build` again to make sure there are no TypeScript errors.
5. Run `npm run start` and play with the plugin in the playground to make sure the plugin still behaves as expected.
6. Run `npm run test` to run any automated tests.
7. Run `npm run lint` to make sure there are no formatting errors.
8. Open a PR with your changes!

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sdthaker**
## The basics

- [x] I [validated my changes](https://developers.google.com/blockly/guides/contribute/samples#making_and_verifying_a_change)

### Resolves

Fixes #2032 

### Proposed Changes

* Enabled strict mode for TS code.

* Fixed TS errors include _(a lot of the errors were repetitive for different pieces of code)_:

   ```Javascript
   TS2345: Argument of type '(evt: KeyboardEvent) => void' is not assignable to parameter of type '(event: Event) => void'.
     Types of parameters 'evt' and 'event' are incompatible.
       Type 'Event' is missing the following properties from type 'KeyboardEvent': altKey, charCode, code, ctrlKey, and 17 
   more.
   ```
   ```Javascript
   TS2531: Object is possibly 'null'.
   ```
   ```Javascript
   TS2345: Argument of type '(e: KeyboardEvent) => void' is not assignable to parameter of type '(event: Event) => void'.
  Types of parameters 'e' and 'event' are incompatible.
    Type 'Event' is not assignable to type 'KeyboardEvent'.
   ```
   ```Javascript
    TS7034: Variable 'topBlockText' implicitly has type 'any[]' in some locations where its type cannot be determined.
   ```
   ```Javascript
   TS7005: Variable 'topBlockText' implicitly has an 'any[]' type.
   ```

### Reason for Changes

To enhance compile-time type checking built-in with TS to avoid potential errors during runtime.

* `npm run lint` results: 
   >image
* `npm run build` results:
   >image
* `npm run test` results:
   >image

### Test Coverage

N/A

### Documentation

N/A

### Additional Information

I'm unsure if what I did here for [L40-43](https://github.com/google/blockly-samples/pull/2079/files#diff-b0dd3df082f16f5f78be79382a91c7d5fab1c2967ea213d2fcc1f38897925330R40-R43) is correct!

### CODE REVIEW

File: plugins/workspace-search/src/workspace_search.ts [hunk 1]
```diff
@@ -156,7 +156,9 @@ export class WorkspaceSearch implements Blockly.IPositionable {
     this.addEvent(this.inputElement, 'input', this, () => this.onInput());
     this.addEvent(this.inputElement, 'click', this, () => {
       this.searchAndHighlight(this.searchText, this.preserveSelected);
-      this.inputElement.select();
+      if (this.inputElement) {
+        this.inputElement.select();
+      }
```
Comment (BeksOmega) [inline plugins/workspace-search/src/workspace_search.ts]: I think you can use an optional chaining operator here.
```suggestion
      this.inputElement?.select();
```

File: plugins/workspace-search/src/workspace_search.ts [hunk 2]
```diff
@@ -195,11 +197,11 @@ export class WorkspaceSearch implements Blockly.IPositionable {
    * @param thisObject The value of 'this' in the function.
    * @param func Function to call when event is triggered.
    */
-  private addEvent(
+  private addEvent<T extends Event>(
     node: Element,
     name: string,
     thisObject: object,
-    func: (event: Event) => void,
+    func: (event: T) => void,
```
Comment (BeksOmega) [inline plugins/workspace-search/src/workspace_search.ts:202]: Woot this looks great!

File: plugins/workspace-search/src/workspace_search.ts [hunk 3]
```diff
@@ -220,7 +222,9 @@ export class WorkspaceSearch implements Blockly.IPositionable {
    */
   addActionBtn(btn: HTMLButtonElement, onClickFn: () => void) {
     this.addBtnListener(btn, onClickFn);
-    this.actionDiv.appendChild(btn);
+    if (this.actionDiv) {
+      this.actionDiv.appendChild(btn);
+    }
```
Comment (BeksOmega) [inline plugins/workspace-search/src/workspace_search.ts]:
```suggestion
    this.actionDiv?.appendChild(btn);
```

File: plugins/workspace-search/src/workspace_search.ts [hunk 4]
```diff
@@ -344,7 +349,10 @@ export class WorkspaceSearch implements Blockly.IPositionable {
    */
   private onInput() {
     if (this.searchOnInput) {
-      const inputValue = this.inputElement.value.trim();
+      let inputValue = '';
+      if (this.inputElement) {
+        inputValue = this.inputElement.value.trim();
+      }
```
Comment (BeksOmega) [inline plugins/workspace-search/src/workspace_search.ts]: Can you cahnge this to `if (this.searchOnInput && this.inputElement)`? Not sure if that works but I think it'd be more readable.

File: plugins/workspace-search/src/workspace_search.ts [hunk 5]
```diff
@@ -437,7 +446,9 @@ export class WorkspaceSearch implements Blockly.IPositionable {
    */
   open() {
     this.setVisible(true);
-    this.inputElement.focus();
+    if (this.inputElement) {
+      this.inputElement.focus();
+    }
```
Comment (BeksOmega) [inline plugins/workspace-search/src/workspace_search.ts]:
```suggestion
    this.inputElement?.focus();
```

**Review by BeksOmega — CHANGES_REQUESTED** [review]
This looks great :D Just a few nits / preferences then LGTM!

**Comment by sdthaker** [comment]
@BeksOmega Thanks for the review! I've pushed the commit! Just wanted to know the reason for using optional chaining for certain pieces of code and not for the others.

**Comment by sdthaker** [comment]
> @BeksOmega Thanks for the review! I've pushed the commit! Just wanted to know the reason for using optional chaining for certain pieces of code and not for the others.

I'm assuming it was a choice about better code readability?

**Comment by BeksOmega** [comment]
> @BeksOmega Thanks for the review! I've pushed the commit! Just wanted to know the reason for using optional chaining for certain pieces of code and not for the others.

So in certain places like assignments, optional chaining is not allowed. For example:

```js
this.myPossiblyNullObj?.property = 'my new value';
```

Because optional chaining makes the expression evaluate to `undefined` if the thing being accessed is null or undefined. So in this case what you're doing is basically:

```js
undefined = 'my new value';
```

Which is not allowed.

And in other cases it is just a style thing like you said :P In general I prefer optional chaining for one-line things. But if you're going to access the object multiple times, I usually prefer to early return.

If you have any questions or preferences about specific cases lmk!

**Review by BeksOmega — APPROVED** [review]

**Comment by sdthaker** [comment]
> > @BeksOmega Thanks for the review! I've pushed the commit! Just wanted to know the reason for using optional chaining for certain pieces of code and not for the others.
> 
> So in certain places like assignments, optional chaining is not allowed. For example:
> 
> ```js
> this.myPossiblyNullObj?.property = 'my new value';
> ```
> 
> Because optional chaining makes the expression evaluate to `undefined` if the thing being accessed is null or undefined. So in this case what you're doing is basically:
> 
> ```js
> undefined = 'my new value';
> ```
> 
> Which is not allowed.
> 
> And in other cases it is just a style thing like you said :P In general I prefer optional chaining for one-line things. But if you're going to access the object multiple times, I usually prefer to early return.
> 
> If you have any questions or preferences about specific cases lmk!

This is great information! Thanks so much! Just wanted to know for my reference to get a better insight on the optional chaining case. :)

**Pull request merged**